---
title: "miRNA-mRNA Interactions in CBF-GLIS"
author: "Jenny Smith"
date: "November 5, 2018"
output: html_document
---

#Set-up

```{r setup}
library(knitr)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),tidy=TRUE, fig.align='center', message = FALSE, fig.width = 5, fig.height = 5)
knitr::opts_knit$set(root.dir = file.path(PROJHOME, '2017.02.15_CBF-GLIS_DEG/2018.03.21_CBF-GLIS_DEGs_Comprehensive/'))
options(stringsAsFactors = FALSE)
```

```{r}
library(stringr)
library(magrittr)
library(dplyr)
library(tidyr)
library(tibble)
library(XML)
getwd()
```

```{r}
# source("~/scripts/RNAseq_Analysis/DifferentialExpn_PathwayAnalysis/DifferentialExpressionPipeline_7.05.18.r")
# source("~/scripts/miRNAseq_analysis/DifferentialExpression_miRNAseq_pipeline_07.24.18.r")
```


#Define Functions to be used

```{r}
corr.miRNA.mRNA <- function(miRNA.Expn, gene.Expn, phenovector,ref){
  library(ggplot2)
  library(psych)
  # miRNA.Expn is a named numeric vector of miRNA log2 RPMs or TMM normalized counts
  #gene.Expn is a named numeric vector of mRNA log2 normalized expn (CPM, TPM, RPKM, etc)
  #phenovector is named characer vector of the groups (eg pos, neg)
  #ref is a charcter vector length 1, which has the group from phenovector that was "control group"
  
  g1 <- names(which(phenovector != ref))
  g2 <- names(which(phenovector == ref))

  #Subset for the group of interest and ensure same order in expression sets
  RPM <- t(miRNA.Expn[,g1])
  GE <- t(gene.Expn[,g1])
  
  #combine the expression sets
  expn.g1 <- RPM %>%
    as.data.frame() %>%
    rownames_to_column("USI") %>%
    inner_join(., 
               rownames_to_column(as.data.frame(GE),"USI"), 
               by="USI")
  
  expn.g2 <- t(miRNA.Expn[,names(phenovector)]) %>%
    as.data.frame() %>%
    rownames_to_column("USI") %>%
    inner_join(., 
               rownames_to_column(as.data.frame(t(gene.Expn[,names(phenovector)])),"USI"), 
               by="USI") %>%
    mutate(Group=phenovector) %>%
    dplyr::select(USI,Group, everything())
    
  #correlation of miRNA to mRNA in group 1, CBF-GLIS in this case. 
  corr <- corr.test(RPM,GE, method="spearman",
                    adjust="BH",ci=FALSE)
  
  #Format the results 
  pearson <- corr$r %>%
    as.data.frame() %>%
    rownames_to_column("MIMAT") %>%
    gather(Gene,SpearmanRho, -MIMAT) 
  
  pvals <- corr$p %>%
    as.data.frame() %>%
    rownames_to_column("MIMAT") %>%
    gather(Gene,Adj.P.val, -MIMAT)
 
  res <- merge(pearson,pvals, by=c("MIMAT","Gene")) %>%
    separate(MIMAT,into = c("Mir","MIMAT"), sep="\\.")
  
  list <- list(expn.g1,expn.g2, res, corr)
  names(list) <- c("Merged_Expression","Merged_Expression_Ref", "Results", "correlation_list")
  
  return(list)
  
}
```


```{r}
queryAnamir <- function(pairs, cutoff=2){
    db <- RMySQL::dbConnect(RMySQL::MySQL(), user = "visitor",
                          password = "visitor",
                          dbname = "visitor",
                          host = "anamir.cgm.ntu.edu.tw")
    
     interaction <- list()
     
     for (i in seq_len(nrow(pairs))){
       # print(i)
       mirna <- pairs[i,1]
       gene <- pairs[i,2]
       corr <- pairs[i,3]

       query <- paste0("SELECT `miRNA_21`, `Gene_symbol`, `Ensembl`,
                      `Gene_ID`, `DIANA_microT_CDS`, `EIMMo`, `Microcosm`,
                      `miRDB`, `miRanda`, `PITA`, `rna22`, `Targetscan`,
                      `Sum`, `miRecords`, `miRTarBase`,
                      `Validate` FROM `all_hsa` WHERE miRNA_21 like '",
                      mirna, "' AND gene_symbol like '", gene, "' ;")

       tmp <- DBI::dbGetQuery(db, query)
  
      #If statement to avoid adding empty query results to the interactions list
       if (nrow(tmp) == 0 && cutoff > 0) {
         next 
       } else {
          # print(colnames(tmp)) #16 columns, #13 is sum and #16 is validate
          tmp <- c(tmp, corr, 0)
          interaction[[i]] <- tmp
       }
     }

    # interaction <- data.frame(do.call(rbind, interaction))
    interaction <- do.call(rbind, interaction)
    
    # disconnect db
    cons <- RMySQL::dbListConnections(RMySQL::MySQL())
    for (con in cons) RMySQL::dbDisconnect(con)

    # add column de novo
    colnames(interaction)[ncol(interaction)] <- c("de novo")
    
    if (cutoff > 1 && nrow(interaction) > 0) {
      del_row <- c()
      for (i in seq_len(nrow(interaction))) {
        if (interaction[i, 13] < cutoff && interaction[i, 16] %in% "FALSE") {
          del_row <- c(del_row, i)
        }
      }
      interaction <- interaction[-(del_row), ]
    }
    
    return(interaction)
}

```

```{r}
extract_core_genes <- function(gage_from_pipeline.res,direction="up", ret.genesets=FALSE){
  
  gage.res.sig <- gage_from_pipeline.res[sapply(gage_from_pipeline.res,length) == 5]

  
  if(direction=="up"){
    idx <- "essSets.Up"
    idx.num <- 4
  }else if (direction=="down"){
    idx <- "essSets.Dn"
    idx.num <- 5
  }
  
  #Remove extraneous gene-sets from the MSigDB C2.All set
  c2.all <- grep("OtherAML_expn_C2.All", names(gage.res.sig))
  i <- sapply(names(gage.res.sig[[c2.all]][[idx]][["coreGeneSets"]]), 
              function(x) grep("^PID_|^REACTOME|^BIOCARTA", x)) 
  i <- unlist(lapply(i, function(x) length(x) > 0))
  
  # gage.res.sig[[c2.all]][[idx]][["coreGeneSets"]] <- gage.res.sig[[c2.all]][[idx]]$coreGeneSets[i]
  
  #select the correct index for the pathway core genesets
  core.genes <- lapply(gage.res.sig, `[[`, idx.num) 
  core.genes <- lapply(core.genes, `[[`, 7) 
  
  if(ret.genesets){
    return(core.genes)
  }
  
  #Define an empty list to append to 
  paths.core.genes <- list()
  #index position for the new lis
  n <- 1
  
  #for loop to un-nest the core.genes, so that they are not in a depth == 2 list
  for (i in 1:length(core.genes)){
    
    gs <- core.genes[[i]]
    
    if(i == 1){
       stop=length(gs)
    }else{
       stop <- n + c(length(gs)-1)
    }
    
    items <- n:stop
    paths.core.genes[items] <- gs
    names(paths.core.genes)[items] <- names(gs)
    
    n <- stop+1
  }

  return(paths.core.genes)
  
}
```

```{r}
find.Pathway.Ints <- function(pathway.genes, Interaction.df, pathway.name=NULL){
  
  Ints <- unlist(lapply(pathway.genes, function(x)
    grep(paste0("^",x,"$"), Interaction.df$Gene_symbol,value = TRUE, ignore.case = TRUE)))
  
  pairs <- Interaction.df %>%
    filter(Gene_symbol %in% Ints) %>%
    mutate(Pair=paste(miRNA_21, Gene_symbol, sep=", ")) %>%
    dplyr::select(Pair)
  
  if(!is.null(pathway.name)){
    colnames(pairs) <- pathway.name
  }

  return(pairs)
  
}
```


#Read in the clinical data

```{r}
CDE <- read.csv("~/reference_mapping-files/TARGET_AML_0531_1031_merged_CDEs_10.29.18.csv")

CDE <- CDE %>%
  filter(!is.na(TARGET.USI.1)) %>%
  mutate(CBFA2T3.GLIS2=gsub("Intermediate", "Yes", CBFA2T3.GLIS2)) %>%
  set_rownames(.$TARGET.USI.1)

dim(CDE)
```




#Read in the mapping-file


```{r}
ID.map <- read.csv(file.path(GENREFS, "miRBase_v21/hsa_gff3_IDMap.csv"))
```



#Read in the oncogene and oncomir reference

```{r}
apop <- read.csv("~/RNA_seq_Analysis/0000.00.02_Reference_GeneInfo/ApoCanD_Database_of_Apoptotic_Protiens.csv")

head(apop)
dim(apop)
```

```{r}
TS <- read.table("~/RNA_seq_Analysis/0000.00.02_Reference_GeneInfo/Human_TumorSuppressorGenes_bioinfo.uth.edu.txt", stringsAsFactors = FALSE, sep="\t", header = TRUE)

head(TS)
# dim(TS) #1217 tumor supressors 
```



#Read in the DEGs and Expression Values. 


```{r}
DEGs <- readRDS("DEGs/TARGET_AML_AllCohorts_CBFGLISvsOtherAML_DEGs_10.23.18.RDS")

CPM <- DEGs$cts.hd.1031$DE$Voom$E
```


```{r}
DEMirs <- get(load("DEMirs/TARGET_AML_1031_CBFGLIS_vs_OtherAML_DEMirs.RData"))

CPM.mir <- DEMirs$DE$Voom$E
```


```{r}
table(CDE[names(DEGs$cts.hd.1031$phenovector), "CBFA2T3.GLIS2"],
      CDE[names(DEGs$cts.hd.1031$phenovector), "RAM.phenotype"] )
```




#Examine the Distribusions

```{r fig.height=4 fig.wid=4}
# range(log2.RPM) # 0.00000 19.20284
# hist(t(log2.RPM))
range(CPM) # -7.952887 21.335325
hist(t(CPM))
```


```{r}
# range(log2.TPM) #0.00000 19.51991
range(CPM.mir) #-4.653424 20.375394
hist(t(CPM.mir))
```



#Quick look at oncogenes and tumor supressors

```{r}
DEGs.1031 <- read.csv("DEGs/TARGET_AML_1031_CBFGLIS_vs_OtherAML_DEGs_updateAnno.csv")
dim(DEGs.1031) #3711 
```

```{r}
DEMirs.1031 <- read.csv("DEMirs/TARGET_AML_1031_CBFGLIS_vs_OtherAML_DEMirs_10.29.18.csv")

dim(DEMirs.1031) #134 
```



#Create Lists of Up and Down Regulated Genes/miRNAs


```{r}
# quantile(DEGs.1031$logFC)
Up.DE <- DEGs.1031 %>%
  filter(logFC > 1) %>%
  dplyr::select("gene") %>%
  unlist() #2,285

Dn.DE <- DEGs.1031 %>%
  filter(logFC < -1 ) %>% 
  dplyr::select("gene") %>%
  unlist() #1,426
```

```{r}
Up.DEMirs <- DEMirs.1031 %>%
  filter(logFC > 1) %>% 
  dplyr::select(gene) %>%
  unlist() #68

Dn.DEMirs <- DEMirs.1031 %>%
  filter(logFC < -1) %>% 
  dplyr::select(gene) %>%
  unlist() #66
```



#Correlation of the Genes to miRNAs


```{r}
samp <- names(DEGs$cts.hd.1031$phenovector)
# samp <- samp[samp != "PAXGKH"] #major outlier in expression. 
length(samp)
```

```{r}
# corrs.UpGenes <- corr.miRNA.mRNA(miRNA.Expn = RPM.fmt[Dn.DEMirs, samp],
#                          gene.Expn = TPMs[Up.DE,samp],
#                          phenovector = DEGs$cts.hd.1031$phenovector[samp],
#                          ref="GroupB")

corrs.UpGenes <- corr.miRNA.mRNA(miRNA.Expn = DEMirs$DE$Voom$E[Dn.DEMirs,],
                         gene.Expn = DEGs$cts.hd.1031$DE$Voom$E[Up.DE,],
                         phenovector = DEGs$cts.hd.1031$phenovector,
                         ref="GroupB")

corrs.UpGenes$Results %>%
  arrange(Adj.P.val) %>%
  head()

# range(corrs.UpGenes$Results$SpearmanRho) # -0.87 to  0.84, why there are still positive correlations?
# write.csv(corrs.UpGenes$Results,"CBFLGIS_UpGenes_DnMirs_AllCorrs.csv", row.names = FALSE)
```


```{r warning=FALSE}
# corrs.DnGenes <- corr.miRNA.mRNA(miRNA.Expn = log2.RPM[Up.DEMirs,samp],
#                                  gene.Expn = log2.TPM[Dn.DE,samp],
#                                  phenovector = DEGs$cts.hd.1031$phenovector[samp],
#                                  ref="GroupB")

corrs.DnGenes <- corr.miRNA.mRNA(miRNA.Expn = DEMirs$DE$Voom$E[Up.DEMirs,],
                                 gene.Expn = DEGs$cts.hd.1031$DE$Voom$E[Dn.DE,],
                                 phenovector = DEGs$cts.hd.1031$phenovector,
                                 ref="GroupB")

corrs.DnGenes$Results %>%
  arrange(Adj.P.val) %>%
#cannot detect relationship bc y does not vary in these genes (all zeros in TPMs in CBFGLIS, down-reg genes)
  filter(!is.na(SpearmanRho)) %>% 
  head()

range(corrs.DnGenes$Results$SpearmanRho,na.rm = T) # -0.90 to 0.89

```



#Examine in detail the correlation matrix 

For cor(), if method is "kendall" or "spearman", Kendall's tau or Spearman's rho statistic is used to estimate a rank-based measure of association. These are more robust and have been recommended if the data do not necessarily come from a bivariate normal distribution.

```{r}
pos <- names(DEGs$cts.hd.1031$phenovector %>% grep("GroupA", ., value=T))
pos
```


```{r}
dim(corrs.UpGenes$correlation_list$r)

head(corrs.UpGenes$correlation_list$r[,1:5])
```


#Update the Mir Annotations

```{r}
UpGenePairs.all <- corrs.UpGenes$Results %>%
  left_join(., dplyr::select(ID.map, Mir.V.21=miR, everything()), by=c("MIMAT"="MIMAT.ID")) 

 # write.csv(UpGenePairs.all,"CBFGLIS_UpGenes_DnMirs_AllCorrs.csv", row.names = FALSE )
```

```{r}
UpGenePairs.all %>%
  filter(Gene=="NCAM1")  %>%
  arrange(Adj.P.val) %>%
  head()
```

```{r}
UpGenePairs <- corrs.UpGenes$Results %>%
  left_join(., dplyr::select(ID.map, Mir.V.21=miR, everything()), by=c("MIMAT"="MIMAT.ID")) %>%
  filter(Adj.P.val < 0.05 & SpearmanRho < 0) %>%
  arrange(Adj.P.val)


head(UpGenePairs) 
dim(UpGenePairs) #452 pairs
```

```{r}
range(UpGenePairs$SpearmanRho) #-0.87 to -0.69
quantile(UpGenePairs$Adj.P.val, probs = seq(0,1,length.out=11))
length(unique(UpGenePairs$Gene)) #304 unqiue genes in miRNA pairs with negative correlation
```

```{r}
DnGenePairs <- corrs.DnGenes$Results %>%
    left_join(., dplyr::select(ID.map, Mir.V.21=miR, everything()), by=c("MIMAT"="MIMAT.ID")) %>%
    filter(Adj.P.val < 0.05 & SpearmanRho < 0) %>%
    arrange(Adj.P.val)

head(DnGenePairs) #13540  pairs
dim(DnGenePairs)
# write.csv(DnGenePairs, "TARGET_AML_CBFGLIS_vs_otherAML_DnGene_UpMirs_antiCorrs.csv", row.names = FALSE)
```

```{r}
range(DnGenePairs$SpearmanRho) #-0.90 -0.53
length(unique(DnGenePairs$Gene)) #1065 unique genes 
```


```{r}
# write.csv(UpGenePairs, "TARGET_AML_1031_CBFGLIS_vs_OtherAML_UpRegDEGs_DnRegDEMirs_Anti-Correlations_RemovedOUTLIER_TMM_7.30.18.csv", row.names = FALSE)
```

```{r}
# write.csv(DnGenePairs, "TARGET_AML_1031_CBFGLIS_vs_OtherAML_DnRegDEGs_UpRegDEMirs_Anti-Correlations_RemovedOUTLIER_TMM_7.30.18.csv", row.names = FALSE)
```



#Query Interaction Databases from anamiR

```{r}
library(anamiR, lib.loc="/home/jlsmith3/R/x86_64-pc-linux-gnu-library/3.5")
```

```{r}
reFormat.anaMir <- function(filt.cor.res){
  cor.mat <- filt.cor.res %>%
        dplyr::select(miRNA=Mir.V.21, Gene, Correlation=SpearmanRho) %>%
        as.matrix()
  return(cor.mat)
}
```

```{r}
up.pairs.fmt <- reFormat.anaMir(UpGenePairs)
dn.pairs.fmt <- reFormat.anaMir(DnGenePairs)


head(up.pairs.fmt)
# dim(up.pairs.fmt) #452 by    3
head(dn.pairs.fmt)
# dim(dn.pairs.fmt) #13540 by 3
```

```{r}
Interaction.Up <- readRDS("miRNA-mRNA_Interactions/anamiR/TARGET_AML_1031_CBFGLIS_vs_OtherAML_UpDEGs_DnMirs.RDS")

# Interaction.Up <- queryAnamir(up.pairs.fmt)

# saveRDS(Interaction.Up, "TARGET_AML_1031_CBFGLIS_vs_OtherAML_UpDEGs_DnMirs.RDS")
dim(Interaction.Up) # 41 18
```


```{r}
Interaction.Dn <- readRDS("miRNA-mRNA_Interactions/anamiR/TARGET_AML_1031_CBFGLIS_vs_OtherAML_DnDEGs_UpMirs.RDS")
# Interaction.Dn <- queryAnamir(dn.pairs.fmt)

# saveRDS(Interaction.Dn, "TARGET_AML_1031_CBFGLIS_vs_OtherAML_DnDEGs_UpMirs.RDS")
dim(Interaction.Dn) #1686  by 18
```


#Query the most up-to-date miRTarBase

The one ine anamiR and multiMir are several years old. 

```{r}
mirTarBase <- read.csv("/fh/fast/meshinchi_s/workingDir/TARGET/Reference_Data/miRBase_v21/miRTarBase_v7.0_hsa_MTI.csv") %>% 
  mutate(Pair=paste(miRNA, Target.Gene, sep="_"))

head(mirTarBase)
dim(mirTarBase) # 502652      by 9
```

```{r}
Int.Up.Clean %>% 
  filter(Validate == "FALSE") %>% 
  mutate(Pair=paste(miRNA_21, Gene_symbol, sep="_")) %>% 
  filter(Pair %in% mirTarBase$Pair)


head(Int.Up.Clean)
```

```{r}
t <- UpGenePairs %>% 
  select()
  mutate(Pair=paste(Mir.V.21, Gene, sep="_")) %>%
  filter(Pair %in% mirTarBase$Pair)
  
```

DnGenePairs

```{r}
Int.Dn.Clean %>% 
  filter(Validate == "FALSE") %>% 
  mutate(Pair=paste(miRNA_21, Gene_symbol, sep="_")) %>% 
  filter(Pair %in% mirTarBase$Pair)
```




#Reformat the Data

```{r}
DEGs.1031.fmt <-DEGs.1031 %>%
  mutate(gene=toupper(as.character(gene)))
```


```{r warning=FALSE}
info.up <-  dplyr::select(UpGenePairs, 
                   miRNA_21=Mir.V.21,
                   Gene_symbol=Gene,
                   MIMAT,
                   Adj.P.val) %>%
  mutate(Gene_symbol=toupper(Gene_symbol))


Int.Up.Clean <- Interaction.Up %>%
  apply(., 2, as.character) %>% #all columns are "list" class, so need this line
  as.data.frame(.) %>%
  
  mutate(Gene_symbol=toupper(Gene_symbol)) %>%
  merge(.,info.up,all.x=TRUE, by=c("miRNA_21", "Gene_symbol")) %>%
  dplyr::select(miRNA_21,MIMAT,Gene_symbol,Ensembl, Gene_ID, Correlation,Adj.P.val, everything()) %>%

  left_join(., dplyr::select(DEGs.1031.fmt,
                      Gene_symbol=gene, 
                      logFC_Gene=logFC,  
                      adj.P.Val_Gene=adj.P.Val), by="Gene_symbol") %>%
  left_join(., dplyr::select(DEMirs.1031,
                      MIMAT, 
                      logFC_miR=logFC,  
                      adj.P.Val_miR=adj.P.Val),by="MIMAT") %>%
  
  mutate_all(funs(ifelse(grepl("^[0-9]|^\\-[0-9]", .), as.numeric(.), .))) %>%
  filter(Validate == TRUE | Sum >= 2) %>%
  
  dplyr::select(miRNA_21, MIMAT,logFC_miR,adj.P.Val_miR, Gene_symbol,logFC_Gene,
                adj.P.Val_Gene,Ensembl,Gene_ID,Correlation,Adj.P.val_Corr=Adj.P.val, everything())
  


head(Int.Up.Clean)
# dim(Int.Up.Clean) #41 by 24
```
 

```{r}
info.dn <-  dplyr::select(DnGenePairs, 
                   miRNA_21=Mir.V.21,
                   Gene_symbol=Gene,
                   MIMAT,Adj.P.val) %>%
  mutate(Gene_symbol=toupper(Gene_symbol))

Int.Dn.Clean <-   Interaction.Dn %>%
  apply(., 2, as.character) %>%
  as.data.frame() %>%
  
  mutate(Gene_symbol=toupper(Gene_symbol)) %>%
  merge(., info.dn, by=c("miRNA_21", "Gene_symbol")) %>%
  dplyr::select(miRNA_21,MIMAT,Gene_symbol,Ensembl, Gene_ID, Correlation,Adj.P.val, everything()) %>%
  
  left_join(., dplyr::select(DEGs.1031.fmt,
                      Gene_symbol=gene, 
                      logFC_Gene=logFC,  
                      adj.P.Val_Gene=adj.P.Val), by="Gene_symbol") %>%
  left_join(., dplyr::select(DEMirs.1031,
                      MIMAT, logFC_miR=logFC, 
                      adj.P.Val_miR=adj.P.Val),by="MIMAT") %>%
  mutate_all(funs(ifelse(grepl("^[0-9]|^\\-[0-9]", .), as.numeric(.), .))) %>%
  filter(Validate == TRUE | Sum >= 2) %>%
  
  dplyr::select(miRNA_21, MIMAT,logFC_miR,adj.P.Val_miR, Gene_symbol,logFC_Gene,
                adj.P.Val_Gene,Ensembl,Gene_ID,Correlation,Adj.P.val_Corr=Adj.P.val, everything())

head(Int.Dn.Clean)
dim(Int.Dn.Clean) #1686   24
```



```{r}
# write.csv(Int.Up.Clean, "TARGET_AML_CBFGLIS_vs_OtherAML_UpRegDEGs_DnRegMIR_miRNA-mRNA_interactions_.csv",
#           row.names = FALSE)
# 
# write.csv(Int.Dn.Clean, "TARGET_AML_CBFGLISvsOtherAML_DnRegDEGs_UpRegMIR_miRNA-mRNA_interactions_.csv",
#           row.names = FALSE)
```



#GSEA with Interacting Genes

```{r}
library(anamiR)
```


```{r}
reformatEnrichment <- function(mat){
  df <- as.data.frame(mat, stringsAsFactors=FALSE) %>%
    set_colnames(gsub(" |\\-", "_", colnames(.))) %>%
    mutate(Raw_P_Value=as.numeric(as.character(Raw_P_Value)),
           Empirical_P_Value=as.numeric(as.character(Empirical_P_Value))) %>%
    mutate(FDR=p.adjust(Empirical_P_Value, method="BH")) %>%
    filter(FDR < 0.05) %>%
    arrange(FDR)
  
  return(df)
}
```

```{r}
Interaction.Up <- readRDS("miRNA-mRNA_Interactions/anamiR/TARGET_AML_1031_CBFGLIS_vs_OtherAML_UpDEGs_DnMirs.RDS")
Interaction.Dn <- readRDS("miRNA-mRNA_Interactions/anamiR/TARGET_AML_1031_CBFGLIS_vs_OtherAML_DnDEGs_UpMirs.RDS")
```

```{r}
#directly from Anamir Github page. 
enrichment_local <- function(
  data_support,
  org = c("hsa", "mmu"),
  per_time = 5000
) {
  org <- match.arg(org)
  db <- RMySQL::dbConnect(RMySQL::MySQL(),
                          user = "visitor",
                          password = "visitor",
                          dbname = "visitor",
                          host = "anamir.cgm.ntu.edu.tw")
  gene_list <- unique(data_support[, 2])
  identified <- length(gene_list)
  if (org %in% "hsa"){
    databases <- c("KEGG_hsa", "Reactome_hsa", "BioCarta_hsa")
    query <- paste0("SELECT DISTINCT Gene_symbol FROM `all_hsa`;")
    gene_all <- DBI::dbGetQuery(db, query)
    total <- length(gene_all[["Gene_symbol"]])
  }
  if (org %in% "mmu"){
    databases <- c("KEGG_mmu", "Reactome_mmu", "MouseCyc_mmu")
    query <- paste0("SELECT DISTINCT gene_symbol FROM `all_mmu`;")
    gene_all <- DBI::dbGetQuery(db, query)
    total <- length(gene_all[["gene_symbol"]])
  }
  pathway_table <- list()
  number <- 1
  for (data in databases) {
    query <- paste0("SELECT DISTINCT term FROM `", data, "`;")
    term_all <- DBI::dbGetQuery(db, query)
    for (i in seq_len(length(term_all[["term"]]))) {
      if (grepl("'", term_all[i, 1])){
        tmp <- strsplit(term_all[i, 1], "'")
        term_all[i, 1] <- paste(tmp[[1]][1], "''", tmp[[1]][2])
      }
      query <- paste0("SELECT * FROM `", data,
                      "` WHERE term LIKE '",
                      term_all[i, 1], "' ;")

      term_info <- DBI::dbGetQuery(db, query)
      gene_set <- length(term_info[["gene_symbol"]])
      term_name <- term_info[1, 2]
      overlap <- length(intersect(term_info[, 1], gene_list))
      if (overlap == 0) {
        next
      }
      percent <- overlap / gene_set
      p <- stats::phyper(overlap - 1,
                    gene_set,
                    total - gene_set,
                    identified,
                    lower.tail = FALSE)
      number <- number + 1
      #permutation
      count <- vector(mode = "numeric", length = per_time + 1)
      permu_time <- per_time
      for (j in seq_len(permu_time)) {
        gene_per <- sample(gene_all[["Gene_symbol"]], identified)
        if (org %in% "mmu"){
          gene_per <- sample(gene_all[["gene_symbol"]], identified)
        }
        overlap_per <- length(intersect(gene_per, term_info[, 1]))
        p_per <- stats::phyper(overlap_per - 1, gene_set,
                        total - gene_set,
                        identified,
                        lower.tail = FALSE)

        count[j] <- p_per
      }
      count[permu_time + 1] <- p
      count <- sort.default(count)
      empirical_p <- which(count == p) / (per_time + 1)
      if (length(empirical_p) > 1) {
        empirical_p <- mean.default(empirical_p)
      }
      pathway_table[[number]] <- c(data, term_name,
                                   gene_set, overlap,
                                   percent, p, empirical_p)
    }
  }
  #disconnect db
  cons <- RMySQL::dbListConnections(RMySQL::MySQL())
  for (con in cons) RMySQL::dbDisconnect(con)
  pathway_table <- do.call(rbind, pathway_table)
  colnames(pathway_table) <- c("Database",
                               "Term",
                               "Total Genes of the Term",
                               "Targets in the Term",
                               "Targets in Total Genes of the Term(%)",
                               "Raw P-Value",
                               "Empirical P-Value")
  return(pathway_table)
}
```

```{r}
up.paths <- enrichment_local(data_support = Interaction.Up, org = "hsa")

# saveRDS(up.paths, "TARGET_AML_1031_CBFGLIS_UpGenes_DnMiRs_Interactions_GSEA.RDS")
```

```{r}
up.paths.fmt <-  reformatEnrichment(up.paths) %>%
  arrange(desc(Targets_in_the_Term), FDR)

head(up.paths.fmt)
dim(up.paths.fmt) #188 signigicant paths
```

```{r}
# write.csv(up.paths.fmt, "TARGET_AML_1031_CBFGLIS_vs_OtherAML_UpDEGs_DnMIR_miRNAInteractions_GSEA.csv")
```

```{r}
dn.paths <- enrichment_local(data_support = Interaction.Dn, org="hsa")

# saveRDS(dn.paths, "TARGET_AML_1031_CBFGLIS_DnGenes_UpMiRs_Interactions_GSEA.RDS")
```

```{r}
dn.paths.fmt <- reformatEnrichment(dn.paths) %>%
    arrange(desc(Targets_in_the_Term), FDR)


head(dn.paths.fmt) 
dim(dn.paths.fmt) #140   8
```

```{r}
# write.csv(dn.paths.fmt, "TARGET_AML_1031_CBFGLIS_vs_OtherAML_DnDEGs_UpMIR_with_miRNAInteractions_GSEA.csv")
```


##GO Term Enrichment

```{r}
library(goseq)
```

```{r}
genes <- rownames(DEGs$cts.hd.1031$DE$eBayesFit$coefficients) %>%
  toupper(.)

v.go.dn <- ifelse(genes %in% Int.Dn.Clean$Gene_symbol, 1,0) %>%
  set_names(genes)

v.go.224_452 <- ifelse(genes %in% Int.Dn.Clean$Gene_symbol[grepl("-224|-425", Int.Dn.Clean$miRNA_21)], 1,0) %>% set_names(genes)

v.go.up <- ifelse(genes %in% Int.Up.Clean$Gene_symbol, 1,0) %>%
  set_names(genes)

table(v.go.dn)
table(v.go.up)
table(v.go.224_452)
```

```{r message=FALSE, warning=FALSE}
supportedOrganisms() %>%
  filter(grepl("hg19", Genome))
```

```{r}
#initial point very close to some inequality constraints?
pwf.dn <- nullp(v.go.dn, "hg19", "geneSymbol")

GO.dn <- goseq(pwf.dn, "hg19", "geneSymbol",test.cats=c("GO:BP")) %>%
  mutate(FDR=p.adjust(over_represented_pvalue, method = "BH")) %>%
  filter(FDR < 0.1)

head(GO.dn)
dim(GO.dn) #166

# write.csv(GO.dn, "CBFGLIS_vs_OtherAML_DnDEGs_UpMIR_with_miRNAInteractions_GO.csv", row.names = FALSE)
```

```{r}
kegg.dn <- goseq(pwf.dn, "hg19","geneSymbol", test.cats = c("KEGG"))  %>%
  mutate(FDR=p.adjust(over_represented_pvalue, method = "BH")) %>%
  filter(FDR < 0.1)


head(kegg.dn)
dim(kegg.dn)
```

```{r}
pwf.224_452 <- nullp(v.go.224_452, "hg19", "geneSymbol")

GO.224_452 <- goseq(pwf.224_452, "hg19", "geneSymbol",test.cats=c("GO:BP")) %>%
  mutate(FDR=p.adjust(over_represented_pvalue, method = "BH")) %>%
  filter(FDR < 0.1)

head(GO.224_452)
# dim(GO.224_452) #39

 # write.csv(GO.224_452, "CBFGLIS_vs_OtherAML_miR224_miR452_with_miRNAInteractions_GO.csv", row.names = FALSE)
```


```{r}
#initial point very close to some inequality constraints  
pwf.up <- nullp(v.go.up, "hg19", "geneSymbol")

GO.up <- goseq(pwf.up, "hg19", "geneSymbol") %>% #test.cats=c("GO:BP")
  mutate(FDR=p.adjust(over_represented_pvalue, method = "BH")) #%>%
  # filter(FDR < 0.1) #None have significant FDR 

head(GO.up)
dim(GO.up)

# range(GO.up$FDR) #non pass multiple hypothesis testing. 
```







#Identify miRNA Target Genes in Pathways of Interest

```{r}
Int.Up.Clean <- read.csv("miRNA-mRNA_Interactions/anamiR/TARGET_AML_CBFGLIS_vs_OtherAML_UpRegDEGs_DnRegMIR_miRNA-mRNA_interactions_.csv")

Int.Dn.Clean <- read.csv("miRNA-mRNA_Interactions/anamiR/TARGET_AML_CBFGLISvsOtherAML_DnRegDEGs_UpRegMIR_miRNA-mRNA_interactions_.csv")

gage.res <- readRDS("GAGE/RData/TARGET_AML_1031_CBFGLIS_vs_OtherAML_GAGE_w.KeggDisease.RDS")
```


```{r}
up.core.sets <- extract_core_genes(gage_from_pipeline.res = gage.res, direction = "up")

up.core.genes <- extract_core_genes(gage_from_pipeline.res = gage.res, direction = "up",ret.genesets = TRUE)
# head(up.core.sets)
length(up.core.sets) #1302
```

```{r}
dn.core.sets <- extract_core_genes(gage_from_pipeline.res = gage.res, direction = "down")
head(dn.core.sets)
length(dn.core.sets) #448
```


```{r}
UpPath.Ints<- lapply(up.core.sets, 
                     find.Pathway.Ints, 
                     Interaction.df=Int.Up.Clean)

UpPath.Ints <- UpPath.Ints[sapply(UpPath.Ints, function(x) nrow(x) >0 )]

length(UpPath.Ints) #582 pathways with
```

```{r}
DnPath.Ints <- lapply(dn.core.sets, 
                      find.Pathway.Ints,
                      Interaction.df=Int.Dn.Clean)

DnPath.Ints <- DnPath.Ints[sapply(DnPath.Ints, function(x) nrow(x) >0 )]

length(DnPath.Ints) #320
```

 
#Examine the Pathways with miRNA interations 

```{r}
UpPaths.Ints.N <- sapply(UpPath.Ints, nrow)

data.frame(Pathway=names(UpPaths.Ints.N), 
           Number.Hits=unlist(UpPaths.Ints.N)) %>%
  arrange(desc(Number.Hits)) %>%
  filter(!grepl("_DN$", Pathway)) %>%
  filter(grepl("TGF|BMP|WNT|Hedge|TNF|ECM|Adherencs|Tight|Platelet|Hippo|mTOR|Rap", Pathway, ignore.case = TRUE))
  # filter(Pathway %in% gage.res$OtherAML_expn_C2.All_11.23.18$essSets.Up$essentialSets)
  # filter(grepl("^hsa", Pathway)) 
  # head(n=50)
```

```{r}
UpPath.Ints[c("KEGG_GAP_JUNCTION","KEGG_HEMATOPOIETIC_CELL_LINEAGE",
              "LABBE_TGFB1_TARGETS_UP","GO_SKIN_DEVELOPMENT",
              "DAVICIONI_TARGETS_OF_PAX_FOXO1_FUSIONS_UP",
              "KINSEY_TARGETS_OF_EWSR1_FLII_FUSION_DN")]
```

```{r}
UpPath.Ints$GO_REGULATION_OF_BMP_SIGNALING_PATHWAY
UpPath.Ints$PID_BMP_PATHWAY
UpPath.Ints$`hsa04310 Wnt signaling pathway`
UpPath.Ints$KEGG_WNT_SIGNALING_PATHWAY
UpPath.Ints$`hsa04512 ECM-receptor interaction`
UpPath.Ints$`hsa04015 Rap1 signaling pathway`
```


Other interesting ideas 
1. EVI1 target genes are enriched. EVI1 is not 
  GPR56 identifies primary human acute myeloid leukemia cells with high repopulating potential in vivo ( doi: https://doi.org/10.1182/blood-2015-11-683649)
  
2. Ikaros target genes are enriched. IKZF3 is upregulared lfc=1.52 above other AMLs
  Signaling Proteins and Transcription Factors in Normal and Malignant Early B Cell Development (https://www.hindawi.com/journals/bmr/2011/502751/)


Possibilities: 
GO_REGULATION_OF_BMP_SIGNALING_PATHWAY
PID_BMP_PATHWAY

UpPath.Ints$`hsa04015 Rap1 signaling pathway`

UpPath.Ints$`hsa04611 Platelet activation`


```{r}
sapply(DnPath.Ints, nrow) %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  set_colnames(c("Pathway","Number.Hits")) %>%
  # arrange(desc(Number.Hits)) %>%
  filter(grepl("Apop|Death|differ|DNA", Pathway, ignore.case = TRUE)) %>%
  # filter(grepl("Hallmark", Pathway, ignore.case = T)) %>%
  # filter(grepl("^GO", Pathway)) %>%
  # filter(grepl("^hsa", Pathway)) %>%
  # filter(grepl("^REACTOME", Pathway)) %>%
  head(n=50)
```

```{r}
# DnPath.Ints[c("GO_B_CELL_ACTIVATION","hsa04612 Antigen processing and presentation")]

DnPath.Ints[c("REACTOME_BMAL1_CLOCK_NPAS2_ACTIVATES_CIRCADIAN_EXPRESSION","HALLMARK_DNA_REPAIR" )]
```



#Plot the Significant anti-correlated genes. 

```{r}
library(ggplot2)
```

```{r}
Int.Up.Clean <- read.csv("miRNA-mRNA_Interactions/anamiR/TARGET_AML_CBFGLIS_vs_OtherAML_UpRegDEGs_DnRegMIR_miRNA-mRNA_interactions_.csv")

Int.Dn.Clean <- read.csv("miRNA-mRNA_Interactions/anamiR/TARGET_AML_CBFGLISvsOtherAML_DnRegDEGs_UpRegMIR_miRNA-mRNA_interactions_.csv")
```



##Melted data frame

```{r}
forPlot <- expand.grid(Int.Up.Clean$miRNA_21, Int.Up.Clean$Gene_symbol, stringsAsFactors=FALSE) %>%
  dplyr::select(miRNA_21=Var1,Gene_symbol=Var2 ) %>%
  full_join(., Int.Up.Clean, by=c("miRNA_21", "Gene_symbol")) %>%
  unique() %>%
  
  group_by(miRNA_21) %>%
  mutate(n.targets=sum(!is.na(logFC_Gene))) %>%
  ungroup() %>%
  
  arrange(desc(n.targets), desc(Gene_symbol)) %>%
  mutate(miR.labs=gsub("hsa-", "", miRNA_21)) %>%
  mutate_at(vars(Sum), funs(ifelse(is.na(.), 0, .))) %>%
  mutate_at(vars(Validate), funs(ifelse(is.na(.), FALSE, .))) %>%
  mutate_at(vars(miR.labs,Gene_symbol), funs(factor(., levels = unique(.))))



head(forPlot)
dim(forPlot)
```

```{r fig.height=7, fig.width=10}
ggplot(forPlot, aes(x=miR.labs, y= Gene_symbol, fill=Correlation, size=Sum)) +
    geom_point(shape=21, stroke=0.1) + #aes(stroke=Validate)
    # ggplot(, aes(x=miR.labs, y= Gene_symbol, fill=Correlation, size=Sum, color=Validate)) +
    geom_point(data = filter(forPlot, Validate==TRUE),
               aes(x=miR.labs, y=Gene_symbol, color=Validate),
               shape=21,stroke=1.65) +
  
  labs(x="", y="", title="miRNA-mRNA Interactions", color="Validated", size="Number of Predictions") +
  theme(plot.margin = margin(t = 3, unit = "mm"),
        plot.title = element_text(hjust = 0.5, size = 18),
                       panel.background = element_rect(fill="white"),
                       panel.grid.major = element_blank(),
                       panel.grid.minor = element_blank(),
                       panel.border = element_rect(color = "black", fill=NA),
                       axis.text = element_text(color = "black"),
                       axis.text.x = element_text(angle = 25,hjust=1,vjust = 1, size = 14),
                       axis.text.y = element_text(size = 10),
                       axis.title = element_text(size = 10),
                       legend.key=element_blank()) +
  scale_fill_gradientn(colors=c("red4","red","orange"), values = c(0,0.75,1)) +
  scale_color_manual(values = c("TRUE"="royalblue1"))


```

```{r}
mir.224_452.shared <- Int.Dn.Clean %>%
  filter(grepl("-224-|-452-", miRNA_21)) %>% 
  # mutate(miRNA=gsub("-3p|-5p","",miRNA_21)) %>%
  arrange(Gene_symbol) %>% 
  group_by(Gene_symbol) %>%
  mutate(shared=ifelse(any(grepl("-224",miRNA_21)) & any(grepl("-452", miRNA_21)),
                       TRUE, FALSE)) %>%
  ungroup() 

only.shared <- mir.224_452.shared %>%
  filter(shared) %>%
  select(miRNA_21, MIMAT,
        -shared,
         everything())

dim(only.shared) 
length(unique(only.shared$Gene_symbol)) #61 
write.csv(only.shared, "TARGET_AML_CBFGLIS_miR_225_452_shared_gene_targets.csv", row.names = FALSE)

only.452 <- mir.224_452.shared %>%
  filter(!shared) %>%
  filter(grepl("-452", miRNA_21)) %>%
  select(miRNA_21, MIMAT,
        -shared,
         everything()) 

length(unique(only.452$Gene_symbol))#66
# write.csv(only.452, "TARGET_AML_CBFGLIS_miR_452_only_gene_targets.csv", row.names = FALSE)



only.224 <- mir.224_452.shared %>%
  filter(!shared) %>%
  filter(grepl("-224", miRNA_21)) %>%
  select(miRNA_21, MIMAT,
        -shared,
         everything())
  
length(unique(only.224$Gene_symbol)) #46
# write.csv(only.224, "TARGET_AML_CBFGLIS_miR_225_only_gene_targets.csv", row.names = FALSE)

```


```{r }
mir.224_452 <- Int.Dn.Clean %>%
  filter(grepl("-224-|-452-", miRNA_21)) 
# dim(mir.224_452) #280  24
# length(unique(mir.224_452$Gene_symbol)) #173

mir.224_452.p <- expand.grid(mir.224_452$miRNA_21, mir.224_452$Gene_symbol, stringsAsFactors=FALSE) %>%
  dplyr::select(miRNA_21=Var1,Gene_symbol=Var2 ) %>%
  full_join(., mir.224_452, by=c("miRNA_21", "Gene_symbol")) %>%
  unique() %>%

  group_by(miRNA_21) %>%
  mutate(n.targets=sum(!is.na(logFC_Gene))) %>%
  ungroup() %>%

  arrange(desc(n.targets), desc(Gene_symbol)) %>%
  mutate(miR.labs=gsub("hsa-", "", miRNA_21)) %>%
  mutate_at(vars(Sum), funs(ifelse(is.na(.), 0, .))) %>%
  mutate_at(vars(Validate), funs(ifelse(is.na(.), FALSE, .))) %>%
  mutate_at(vars(miR.labs,Gene_symbol), funs(factor(., levels = unique(.))))

head(mir.224_452.p)
dim(mir.224_452.p)
```

This doesnt make a good figure because the dimensions are just too rectangular, and 173 genes is WAY to many to put into a main figure. readers will not be able to scan that effectively. 

```{r}
length(unique(mir.224_452.p$miRNA_21)) #4 miRs
length(unique(mir.224_452.p$Gene_symbol)) #173 genes
```

```{r fig.hieght=5, fig.width=20}
ggplot(mir.224_452.p, aes(x= Gene_symbol, y=miR.labs , fill=Correlation, size=Sum)) +
    geom_point(shape=21, stroke=0.1) + #aes(stroke=Validate)
    geom_point(data = filter(mir.224_452.p, Validate==TRUE),
               aes(x=Gene_symbol, y= miR.labs, color=Validate),
               shape=21,stroke=1.65) +
  
  labs(x="", y="", title="miRNA-mRNA Interactions: \n Up-Regulated miRNAs vs Down-Regulated Genes", 
       color="Validated", size="Number of Predictions") +
  
  theme(plot.margin = margin(t = 3, unit = "mm"),
        plot.title = element_text(hjust = 0.5, size = 20),
                       panel.background = element_rect(fill="white"),
                       panel.grid.major = element_blank(),
                       panel.grid.minor = element_blank(),
                       panel.border = element_rect(color = "black", fill=NA),
                       axis.text = element_text(color = "black"),
                       axis.text.x = element_text(angle = 35,hjust=1,vjust = 1, size = 6, face="bold"),
                       axis.text.y = element_text(size = 15, face="bold"),
                       axis.title = element_text(size = 10),
                       legend.key=element_blank(), 
                      legend.text = element_text(size=18)) +
  scale_fill_gradientn(colors=c("red4","red","orange"),
                       values = c(0,0.75,1)) + 
  
  scale_size_area(max_size = 6) +
  scale_color_manual(values = c("TRUE"="blue"))

```


## Apoptotic Genes 

```{r}
tumorSuppressors <- read.csv("~/RNA_seq_Analysis/0000.00.02_Reference_GeneInfo/TumorSuppressor_Oncogenes_Kinase_Genelist_Cell_Reports_LiDing_2018.csv")

head(tumorSuppressors)
```

```{r}
apop.ints <- Int.Dn.Clean %>%
  # filter(Gene_symbol %in% tumorSuppressors$TumorSuppressor)
  filter(Gene_symbol %in% apop$Protein | Gene_symbol %in% TS$GeneSymbol |
           grepl("DBP$|PER2|GLI", Gene_symbol) | Gene_symbol %in% tumorSuppressors$TumorSuppressor)
  # filter(grepl("GLI|TNFR|CASP1|PER2", Gene_symbol)) #35 interactions

apop.ints.p <- expand.grid(apop.ints$miRNA_21, apop.ints$Gene_symbol,
                           stringsAsFactors=FALSE) %>%
  dplyr::select(miRNA_21=Var1,Gene_symbol=Var2 ) %>%
  
  full_join(., apop.ints, by=c("miRNA_21", "Gene_symbol")) %>%
  mutate_at(vars(Correlation), funs(round(., digits = 2))) %>%
  unique() %>%

  group_by(miRNA_21) %>%
  mutate(n.targets=sum(!is.na(logFC_Gene))) %>%
  ungroup() %>%

  arrange(Gene_symbol) %>%
  # arrange(Correlation,desc(n.targets), desc(Sum)) %>% #desc(Gene_symbol) desc(n.targets)
  mutate(miR.labs=gsub("hsa-", "", miRNA_21)) %>%
  mutate_at(vars(Sum), funs(ifelse(is.na(.), 0, .))) %>%
  mutate_at(vars(Validate), funs(ifelse(is.na(.) | Sum==0, FALSE, .))) %>%
  
  dplyr::select(miRNA_21, Gene_symbol,Correlation,n.targets, Validate,Sum,miR.labs) %>%
  
  mutate_at(vars(Gene_symbol),
            funs(factor(., levels = rev(unique(.)) ))) %>%
  mutate_at(vars(miR.labs),
            funs(factor(., levels = unique(.) )))



head(apop.ints.p)
# tail(apop.ints.p, n=20)
# dim(apop.ints.p)
```

```{r}
length(unique(apop.ints.p$miRNA_21)) # 36
length(unique(apop.ints.p$Gene_symbol)) # 38

# saveRDS(apop.ints.p, "miRNA-mRNA_Interactions/apoptosis_genes_interactions.RDS")
# write.csv(apop.ints.p,"miRNA-mRNA_Interactions/apoptosis_genes_interactions.csv", row.names = FALSE)
```

```{r fig.height=16, fig.width=24}
# dev.off()

# tiff("Apop_TumorSuppressor_DnGenes_UpMirs_DotPlot.tiff", height = 16, width = 24, units="in", res=300)
ggplot(apop.ints.p, aes(x=miR.labs, y= Gene_symbol, fill=Correlation, size=Sum)) +
    geom_point(shape=21, stroke=0.1) + #aes(stroke=Validate)
     geom_point(data = filter(apop.ints.p, Validate==TRUE),
               aes(x=miR.labs, y=Gene_symbol, color=Validate),
               shape=21,stroke=1.65) +
  
  labs(x="", y="", title="miRNA-mRNA Interactions: Up-Regulated miRNAs vs Down-Regulated Genes",
       color="Validated", size="Number of Predictions") +
  
  theme(plot.margin = margin(t = 3, unit = "mm"),
        plot.title = element_text(hjust = 0.5, size = 30, vjust=2),
                       panel.background = element_rect(fill="white"),
                       panel.grid.major = element_blank(),
                       panel.grid.minor = element_blank(),
                       panel.border = element_rect(color = "black", fill=NA),
                       axis.text = element_text(color = "black"),
                       axis.text.x = element_text(angle = 35,hjust=1,vjust = 1, size = 22, face="bold"),
                       axis.text.y = element_text(size = 22, face="bold"),
                       axis.title = element_text(size = 10),
                       legend.key=element_blank(), 
                       legend.text = element_text(size=18), 
                       legend.title = element_text(size=22)) +
  
  scale_fill_gradientn(colors=c("red4","red","orange"), values = c(0,0.75,1)) + 
  scale_size_area(max_size = 15.0) +
  scale_color_manual(values = c("TRUE"="blue"))

# dev.off()
```



#Scatter Plots

```{r}
ggscatter_custom <- function(corr.miRNA.mRNA.res, pairs.of.interest,data=FALSE,single.group="red",el=FALSE,m.p=FALSE, npc.x=1, npc.y=1){
  library(ggpubr)
  
  g1.dat <- corr.miRNA.mRNA.res$Merged_Expression  %>%
    set_colnames(., gsub("^hsa.+\\.(M.+)", "\\1",colnames(.), ignore.case = TRUE)) %>%
    mutate(GROUP=rep(single.group, nrow(.))) %>%
    dplyr::select(USI,GROUP,pairs.of.interest$MIMAT, pairs.of.interest$Gene_symbol)
  
  g2.dat <- corr.miRNA.mRNA.res$Merged_Expression_Ref %>%
    set_colnames(., gsub("^hsa.+\\.(M.+)", "\\1",colnames(.), ignore.case = TRUE)) %>%
    dplyr::select(USI,GROUP,pairs.of.interest$MIMAT, pairs.of.interest$Gene_symbol)

  if(data){
    return(g2.dat)
  }
  
  
  x.pos <- min(g1.dat[,pairs.of.interest$Gene_symbol])
  y.pos <- min(g1.dat[,pairs.of.interest$MIMAT] - 1)
  
 g1.plot <-  ggscatter(data = g1.dat,
          y = pairs.of.interest$MIMAT,
          x=pairs.of.interest$Gene_symbol,
          color=single.group,
          fill=single.group,
          title=paste( pairs.of.interest$Gene_symbol, "vs", "\n", pairs.of.interest$miRNA_21),
          xlab="Log2( normalized mRNA + 1)",
          ylab="Log2( normalized miRNA + 1)",
          size=5,
          cor.coef=TRUE,
          cor.coeff.args = list(method = "spearman",
                                label.x=x.pos,
                                label.y=y.pos,
                                output.type= "text",
                                label.sep = "\n",
                                size=7.5, 
                                fontface = "bold"),
          fullrange = TRUE,
          ellipse.level = 0.95,
          ellipse = el,
          ellipse.alpha=0.2,
          ellipse.type = "t",
          mean.point = m.p, 
          ylim=c(y.pos-0.2,max(g1.dat[,pairs.of.interest$MIMAT]))) +
  geom_smooth(method = "lm",alpha=0.2, color="black") +
  theme(plot.title = element_text(hjust = 0.5, size = 24),
                       panel.background = element_rect(fill="white"),
                       panel.grid.major = element_blank(),
                       panel.grid.minor = element_blank(),
                       panel.border = element_rect(color = "black", fill=NA),
                       axis.text = element_text(color = "black"),
                       axis.text.x = element_text(angle = 0,hjust=0.5,vjust = 0.5, size = 24),
                       axis.text.y = element_text(size = 24),
                       axis.title = element_text(size = 16))

# all.plot <-  ggplot(g2.dat,
#                     aes_string(y=pairs.of.interest$MIMAT, x=pairs.of.interest$Gene_symbol,color="Group")) +
#     geom_point(alpha=0.65) +
#     theme_bw() +
#     geom_smooth(method = "lm") +
#     scale_color_manual(values=c("GroupA"="red", "GroupB"="blue3")) +
#     labs(x="Log2( TPM + 1)",
#          y="Log2( RPM + 1)",
#          title=paste( pairs.of.interest$Gene_symbol, "vs", pairs.of.interest$miRNA_21)) +
#     geom_point(data = subset(g2.dat, Group=="GroupA"),
#              aes_string(y=pairs.of.interest$MIMAT, x=pairs.of.interest$Gene_symbol), size=2.5,alpha=1) +
#   theme_numX
# 
# 
#   res <- list("g1.plot"=g1.plot, "all.plot"=all.plot)
#   return(res)
  
return(g1.plot)
}
```




##Upregulated Genes with TARGETS 

https://bioconductor.org/packages/release/bioc/vignettes/anamiR/inst/doc/IntroductionToanamiR.html#data-source

From output, column Sum tells us the total hits by 8 predict databases and column Validate tells us if this interaction have been validated.


```{r}
library(gridExtra)
```

```{r}
pairs.of.interest.up <- Int.Up.Clean %>%
  # filter(Validate==TRUE) %>%
  # filter(Gene_symbol %in% onco$OncogeneName) %>%
  # filter( grepl("GLIS2|NCAM1|PTCH1", Gene_symbol)) %>%
  arrange(Gene_symbol)

# pairs.of.interest.up
```

```{r}
colnames(corrs.UpGenes$Merged_Expression) <-  toupper(colnames(corrs.UpGenes$Merged_Expression)) %>%
  gsub("-", "\\.", .)

colnames(corrs.UpGenes$Merged_Expression_Ref) <-  toupper(colnames(corrs.UpGenes$Merged_Expression_Ref)) %>%
    gsub("-", "\\.", .)
```

```{r}
Up.Targets <- pairs.of.interest.up %>%
  rowwise() %>%
  do(plots=ggscatter_custom(corr.miRNA.mRNA.res = corrs.UpGenes, ., single.group = "red"))

Up.Targets$Comparisons <- paste0(gsub("-", "\\.", pairs.of.interest.up$miRNA_21), "_", pairs.of.interest.up$Gene_symbol)
names(Up.Targets$plots) <- Up.Targets$Comparisons

# Up.Targets$plots
```

```{r}
UpPath.Ints$GO_REGULATION_OF_BMP_SIGNALING_PATHWAY
UpPath.Ints$PID_BMP_PATHWAY
```

```{r}
Up.Targets$plots$hsa.miR.222.3p_RGMB
Up.Targets$plots$hsa.let.7b.3p_SOX11
```


```{r fig.width=12.5, fig.height=10}
ml <- marrangeGrob(Up.Targets$plots[1:6], nrow=2, ncol=3)
ml
# ggsave("TARGET_AML_1031_CBFGLIS_vs_OtherAML_UpGenes_DnMirs.pdf", ml, device = "pdf", height = 10, width = 12.5)
```

```{r fig.height=5, fig.width=16.5}
p1 <- duplicate(Up.Targets$plots$hsa.miR.196b.5p_ITGA2) +
  scale_y_continuous(breaks = c(4,6,8,10)) +
  scale_x_continuous(breaks = c(1:7))
p1$coordinates$limits$y <- c(3.5,10)
p1$coordinates$limits$x <- c(1.2,6.5)

p2 <- duplicate(Up.Targets$plots$hsa.miR.196b.5p_DNMT3B) +
  scale_y_continuous(breaks = c(4,6,8,10)) +
  scale_x_continuous(breaks = c(1:7))
p2$coordinates$limits$y <- c(3.5,10)
p2$coordinates$limits$x <- c(2.9,7)

p3 <- duplicate(Up.Targets$plots$hsa.miR.199a.3p_SLITRK5) + 
  scale_y_continuous(breaks = c(4,6,8,10)) +
  scale_x_continuous(breaks = c(1:7))

p3$coordinates$limits$y <- c(3.5,10)
p3$coordinates$limits$x <- c(2.6,6.5)

p4 <- duplicate(Up.Targets$plots$hsa.miR.199b.3p_SLITRK5) +
  scale_y_continuous(breaks = c(4,6,8,10)) +
  scale_x_continuous(breaks = c(1:7))
p4$coordinates$limits$y <- c(3.5,10)
p4$coordinates$limits$x <- c(2.6,6.5)

GOI.Up <- arrangeGrob(p1,p2,p3,p4,
            nrow = 1, ncol = 4)

plot(GOI.Up)
# saveRDS(GOI.Up, "CBFGLIS_UpGenes_DownMirs_Scatterplots.RDS")
# ggsave("TARGET_AML_1031_CBFGLIS_vs_OtherAML_GOI.Up_Scatterplots.tiff", GOI.Up, device = "tiff", height = 5, width = 17,dpi = 600)
```


## Down-Regulate Genes with miRNA Targets 

```{r}
POI.TS <- Int.Dn.Clean %>%
  arrange(desc(logFC_miR),Adj.P.val_Corr) %>% 
  filter(grepl("GLI|TNFR|CASP1|PER2", Gene_symbol))
  # filter(grepl("miR-181|miR-130", miRNA_21))

# head(pairs.of.interest.dn)
dim(POI.TS) #35
# write.csv(POI.TS, "TS_MiRNAs_Ints.csv", row.names = FALSE)
```

```{r}
table(POI.TS$Gene_symbol)
```

```{r}
POI.highMirs <- Int.Dn.Clean %>%
  arrange(desc(logFC_miR),Adj.P.val_Corr, desc(logFC_Gene)) %>% 
  filter(grepl("miR-224-5p|miR-224-3p|miR-452-5p|miR-452-3p|miR-6507-5p,", miRNA_21))
  # filter(grepl("miR-181|miR-130", miRNA_21))

# POI.highMirs
dim(POI.highMirs) #280 - still too many for the figures... 
# write.csv(POI.highMirs, "CBFGLIS_UpMir_DnGenes.csv", row.names = FALSE)
```


```{r}
POI.224 <- POI.highMirs %>%
  filter(grepl("224", miRNA_21)) 

POI.452 <- POI.highMirs %>%
  filter(grepl("452", miRNA_21)) 


v <- venn.diagram(x=list("miR.224"=POI.224$Gene_symbol, "miR.452"=POI.452$Gene_symbol), filename = "miR.224_miR.452.tiff")

intersect(POI.224$Gene_symbol, POI.452$Gene_symbol) #61 in common
setdiff(POI.224$Gene_symbol, POI.452$Gene_symbol) 
```


```{r}
POI.RAM <- Int.Dn.Clean %>%
  arrange(desc(logFC_miR),Adj.P.val_Corr, desc(logFC_Gene)) %>% 
  filter(grepl("HLA-DR|^PTPRC$|CD38", Gene_symbol)) %>%
  mutate(Gene_symbol=gsub("-", "\\.", Gene_symbol))

dim(POI.RAM) #16
# write.csv(POI.RAM, "CBFGLIS_UpMir_DnGenes_RAM.csv", row.names = FALSE)
```


```{r}
colnames(corrs.DnGenes$Merged_Expression) <-  toupper(colnames(corrs.DnGenes$Merged_Expression)) %>%
  gsub("-", "\\.", .)

colnames(corrs.DnGenes$Merged_Expression_Ref) <-  toupper(colnames(corrs.DnGenes$Merged_Expression_Ref)) %>%
  gsub("-", "\\.", .)

```

```{r fig.width=12.5, fig.height=10}
Dn.Apop.TS <- POI.TS %>%
  rowwise() %>%
  do(plots=ggscatter_custom(corr.miRNA.mRNA.res = corrs.DnGenes, ., single.group = "red")) 

Dn.Apop.TS$Comparisons <- paste0(gsub("-","\\.",POI.TS$miRNA_21),"_", POI.TS$Gene_symbol)
names(Dn.Apop.TS$plots) <- Dn.Apop.TS$Comparisons

m_DnGenes <- marrangeGrob(grobs = Dn.Apop.TS$plots, ncol=3, nrow = 2 )

m_DnGenes
# ggsave("TARGET_AML_1031_CBFGLIS_vs_OtherAML_ApopGenes_Scatterplot.pdf", m_DnGenes, device = "pdf", height = 10, width = 12.5)
```



```{r fig.width=12.5, fig.height=8}
Dn.RAM <- POI.RAM %>%
  rowwise() %>%
  do(plots=ggscatter_custom(corr.miRNA.mRNA.res = corrs.DnGenes, ., single.group = "red")) 

Dn.RAM$Comparisons <- paste0(gsub("-","\\.",POI.RAM$miRNA_21),"_", POI.RAM$Gene_symbol)
names(Dn.RAM$plots) <- POI.RAM$Comparisons

m_RAM <- marrangeGrob(grobs = Dn.RAM$plots, ncol=3, nrow = 2 )

m_RAM
# ggsave("TARGET_AML_1031_CBFGLIS_vs_OtherAML_RAMgenes_Scatterplot.pdf", m_RAM, device = "pdf", height = 10, width = 12.5)
```


```{r fig.height=5, fig.width=17}
p1 <- duplicate(Dn.Apop.TS$plots$hsa.miR.181b.5p_CASP1) +
  scale_y_continuous(breaks = c(10,12,14,16)) +
  scale_x_continuous(breaks = c(0:6))
p1$coordinates$limits$y <- c(10.1,16)
p1$coordinates$limits$x <- c(0,6.2)


p2 <- duplicate(Dn.Apop.TS$plots$hsa.miR.181b.5p_PER2) + 
  scale_y_continuous(breaks = c(10,12,14,16)) +
  scale_x_continuous(breaks = c(0:6))

p2$coordinates$limits$y <- c(10.1,16)
p2$coordinates$limits$x <- c(3.1, 6.5)


p3 <- duplicate(Dn.Apop.TS$plots$hsa.miR.130a.3p_GLIPR1) +
  scale_y_continuous(breaks = c(4,6,8,10,12,14,16)) +
  scale_x_continuous(breaks = c(0:6))
p3$coordinates$limits$y <- c(4,16)
p3$coordinates$limits$x <- c(-0.1,6.5)


p4 <- duplicate(Dn.Apop.TS$plots$hsa.miR.425.5p_TNFRSF10B) +
  scale_y_continuous(breaks = c(6,8,10,12,14,16)) +
  scale_x_continuous(breaks = c(0:6))
p4$coordinates$limits$y <- c(8.4,16)
p4$coordinates$limits$x <- c(2.5,6.2)

GOI.Dn <- arrangeGrob(p1,p2,p3,p4,
            nrow = 1, ncol = 4)

# saveRDS(GOI.Dn, "CBFGLIS_DnGenes_UpMirs_Scatterplots.RDS")
plot(GOI.Dn)
```



https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4826477/

Core circadian clock genes regulate leukemia stem cells in AML

Circadian Disruption Induces Myeloid Differentiation and Impaired Cell Cycle Progression
We next examined the consequences of acute disruption of the circadian rhythm in AML using a small molecule. SR9011 is a potent and specific agonist of the nuclear hormone receptors RevErb α and Rev-Erb β, which inhibit Bmal1 transcription via a feedback loop (Burris et al., 2013; Solt et al., 2012). Treatment of quaternary transplant MLL-AF9 cells with SR9011 for 3 days in vitro resulted in a dose-dependent reduction in cell viability (EC50 = 1.8 μM) (Figure 2B). In agreement with known roles of Rev-Erb α and Rev-Erb β within the circadian pathway, short duration treatment of leukemia cells with the drug resulted in decreased expression of Bmal1 and its target gene Per2 (Figure 2C). In addition, we observed dose-dependent suppression of key regulators LSC self-renewal (HoxA9, Meis1, Ctnnb1), indicating that the circadian rhythm circuitry influences functions related to stemness (Figure 2D).


Down regulation of circadian clock gene Period 2 accelerates breast cancer growth by altering its daily growth rhythm
https://link.springer.com/article/10.1007%2Fs10549-008-0133-z



#Barplot of Enriched Paths 

```{r}
dn.paths.fmt <- read.csv("mRNA-miRNA_Interactions/TARGET_AML_1031_DnRegDEGs_with_miRNAInteractions_GSEA.csv", stringsAsFactors = FALSE, row.names = 1)

head(dn.paths.fmt)
```

```{r}
sel.dnPaths <- dn.paths.fmt %>%
  filter(Total_Genes_of_the_Term > 15 & Total_Genes_of_the_Term < 500) %>%
  mutate(FDR.Raw = p.adjust(Raw_P_Value, method = "BH")) %>%
  # filter(grepl("KEGG", Database)) %>%
  arrange(FDR.Raw) %>%
  dplyr::select(Database,Term, FDR.Raw) %>%
  mutate(Database=gsub("_hsa","",Database)) %>%
  mutate(path=gsub("_| ", "\n", paste(Database, Term))) %>%
  slice(1:5)

sel.dnPaths
```

```{r}
tiff("CBFGLIS_vs_otherAML_mRNA-miRNA_interactions_Gsea_dnDEGs_barplot.tiff", height= 7, width = 7, units="in", res=600)

ggplot(sel.dnPaths, aes(y=-log2(FDR.Raw),x=reorder(sel.dnPaths$path, -sel.dnPaths$FDR.Raw), fill=Term)) +
  geom_bar(stat="identity") +
  scale_y_continuous(limits = c(0,16),breaks = seq(0,16,by=2)) +
  scale_fill_brewer(palette="Dark2") +
  geom_hline(aes(yintercept = -log2(0.05)), linetype="dashed") +
  theme_numX + labs(x="", y="-log2(FDR)") + guides(fill=FALSE) +
  coord_flip()

dev.off()
```


```{r}
up.paths.fmt <- read.csv("mRNA-miRNA_Interactions/TARGET_AML_1031_UpRegDEGs_with_miRNAInteractions_GSEA.csv", stringsAsFactors = FALSE)

head(up.paths.fmt)
```

```{r}
sel.upPaths <- up.paths.fmt %>%
  filter(Total_Genes_of_the_Term > 15 & Total_Genes_of_the_Term < 500) %>%
  mutate(FDR.Raw = p.adjust(Raw_P_Value, method = "BH")) %>%
  # filter(grepl("KEGG", Database)) %>%
  arrange(FDR.Raw) %>%
  dplyr::select(Database,Term, FDR.Raw) %>%
  mutate(Database=gsub("_hsa","",Database)) %>%
  mutate(path=gsub("_| ", "\n", paste(Database, Term))) %>%
  slice(1:5)


sel.upPaths
```


```{r}
tiff("CBFGLIS_vs_otherAML_mRNA-miRNA_interactions_Gsea_UpDEGs_barplot.tiff", height= 7, width = 7, units="in", res=600)

ggplot(sel.upPaths, aes(y=-log2(FDR.Raw),x=reorder(sel.upPaths$path, -sel.upPaths$FDR.Raw), fill=path)) +
  geom_bar(stat="identity") +
  scale_y_continuous(limits = c(0,16),breaks = seq(0,16,by=2)) +
  scale_fill_brewer(palette="Dark2") +
  geom_hline(aes(yintercept = -log2(0.05)), linetype="dashed") +
  theme_numX +
  labs(x="", y="-log2(FDR)") + guides(fill=FALSE) +
  coord_flip()

dev.off()
```




#Session Info

```{r}
sessionInfo()
```






















