---
title: 'DE Mirs Workflow with CBFA2T3-GLIS2'
author: "Jenny Smith"
date: "July 25, 2018"
output: html_document
---

#Set-up

```{r setup}
library(knitr)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),tidy=TRUE, fig.align='center', fig.width = 10, fig.height = 10)
knitr::opts_knit$set(root.dir = file.path(PROJHOME,"2017.02.15_CBF-GLIS_DEG/2018.03.21_CBF-GLIS_DEGs_Comprehensive/"))
options(stringsAsFactors = FALSE)

```


```{r message = FALSE, warning=FALSE}
library(stringr)
library(magrittr)
library(ggplot2)
library(dplyr)
library(tibble)
library(tidyr)
library(readr)
library(gtools)
library(Cairo)
getwd()
```


```{r}
source(file.path(SCRIPTS,"miRNAseq_analysis/DifferentialExpression_miRNAseq_pipeline_07.24.18.r"))
```


*Optimization of miRNA-seq data preprocessing*
Shirley Tam,  Ming-Sound Tsao,  John D. McPherson
Briefings in Bioinformatics, Volume 16, Issue 6, 1 November 2015, Pages 950–963, https://doi.org/10.1093/bib/bbv019


*Genome-wide analysis of thapsigargin-induced microRNAs and their targets in NIH3T3 cells*
Jody Groenendyk,a Xiao Fan,b Zhenling Peng,b Yaroslav Ilnytskyy,c Lukasz Kurgan,b and Marek Michalaka,⁎
Genom Data. 2014 Dec; 2: 325–327.
Published online 2014 Oct 7. doi:  10.1016/j.gdata.2014.10.002


#Read in the clinical Data

```{r}
# merged <- read.csv("~/reference_mapping-files/00_Old/TARGET_AML_0531_1031_merged_CDEs_3.01.19.csv", 
#                    stringsAsFactors = FALSE)
merged <- read.csv(file.path(CDE,"Merged/TARGET_AML_0531_1031_merged_CDEs_9.18.20.csv"))

merged <- merged %>%
  filter(!is.na(USI), USI != "Unknown") %>%
  mutate(CBFA2T3.GLIS2=case_when(
    grepl("CBFA2T3.GLIS2", Primary.Fusion) ~ "Yes",
    ScreenedForFusion == "No" ~ "Unnknown",
    TRUE ~ "No")) %>% 
  # mutate_at(vars(CBFA2T3.GLIS2), 
  #           ~ifelse(CBFA2T3.GLIS2=="Intermediate", "Yes", .)) %>%
  set_rownames(.$USI)

dim(merged)
head(merged[,1:5])
```
 
```{r}
table(merged$CBFA2T3.GLIS2) #40
```

 

#Read in the miRNAseq Expression Sets

```{r}
ID.map <- read_csv(file.path(GENREFS,"miRBase_v21/hsa_gff3_IDMap.csv"))

head(ID.map)
```

```{r}
RPM.1031.orig <- read_csv(file.path(TARGET,"RNA/miRNAseq/level3/mature_miRNA/2017July_BCCA_1031_miRBase20_Illumina_data/TARGET_AML_AAML1031_expn_matrix_mimat_miRNA_RPM.csv"))

head(RPM.1031.orig[,1:5])
```

```{r}
mir.cts.1031 <- read_csv(file.path(TARGET,"RNA/miRNAseq/level3/mature_miRNA/2017July_BCCA_1031_miRBase20_Illumina_data/TARGET_AML_AAML1031_expn_matrix_mimat_miRNA_rawCounts.csv")) 

head(mir.cts.1031[,1:5])
```


#Reformat the counts

```{r}
reformat.mirs <- function(df,cell.lines=FALSE){  
  df <- df %>%
    rename_at(vars(contains("BCCA")),
              function(x) gsub("^.+(Kas|MV4.+R)","\\1", x) ) %>%
    rename_all(function(x) str_split_fixed(x, "-", n=2)[,1])  
  
    if(cell.lines){
      df <- select(df, mir, matches("^Kas|MV4")) 
    }else{
      df <- select(df, -matches("^Kas|MV4"))
    }
  
  df <- column_to_rownames(df,"mir")

  
}
```

```{r}
RPM.1031 <- reformat.mirs(RPM.1031.orig)

head(RPM.1031[,1:5])
dim(RPM.1031) # 2576 1135
```

```{r}
RPM.cells <- reformat.mirs(RPM.1031.orig, cell.lines = TRUE)
head(RPM.cells)
```

```{r}
mir.cts.1031 <- reformat.mirs(mir.cts.1031)

head(mir.cts.1031[,1:5])
# dim(mir.cts.1031) #2576 by 1135
```



#Differential miRNAs Analysis 

##AML

```{r}
pos.all <- merged[merged$CBFA2T3.GLIS2 == "Yes", 1]
neg.all <- merged[merged$CBFA2T3.GLIS2 == "No", 1]


pos.1031 <- intersect(colnames(mir.cts.1031), pos.all) #24 between 1031 and a handful of  0531
neg.1031 <- intersect(colnames(mir.cts.1031), neg.all) #1025 1031 negatives
```


```{r message=FALSE}
# tiff("TARGET_AML_1031_CBGLIS_vs_OtherAML_DEMirs_Heatmap_10.29.18.tiff", width = 10, height = 10, units="in", res=600)

DEMirs <- twoGroups_DEMirs(expnData = mir.cts.1031, 
                           clinData = merged, 
                           col = "CBFA2T3.GLIS2", 
                           ref = "No",
                           RPM = FALSE,
                           BM = FALSE)
# dev.off()
```


```{r}
# save(DEMirs, file="TARGET_AML_1031_CBFGLIS_vs_OtherAML_DEMirs.RData")
DEMirs <- get(load("DEMirs/TARGET_AML_1031_CBFGLIS_vs_OtherAML_DEMirs.RData"))
table(DEMirs$phenovector)
```


```{r}
# summary(DEMirs)
dim(DEMirs$DE$Voom$E) #684  mirs considered expressed. 
head(DEMirs$DE$Voom$E[,1:5])
```


```{r}
DEMirs.df <- extract_DEMirs(DEMirs)

DEMirs.df <- DEMirs.df %>%
  mutate(miR.v20=str_split_fixed(gene, "\\.", n=2)[,1], 
         MIMAT=str_split_fixed(gene, "\\.", n=2)[,2]) %>%
  left_join(., ID.map, by=c("MIMAT"="MIMAT.ID")) %>%
  dplyr::select(miR,miR.v20,MIMAT, everything(), gene)


head(DEMirs.df)
# dim(DEMirs.df)
# write.csv(DEMirs.df, "TARGET_AML_1031_CBFGLIS_vs_OtherAML_DEMirs_10.29.18.csv", row.names = FALSE)
```


```{r}
highly.expressed <- DEMirs.df %>% 
  filter(logFC >= quantile(logFC)[4])

dim(highly.expressed)

highly.Expressed.TMM <- DEMirs$DE$Voom$E[highly.expressed$gene,
                                         names(DEMirs$phenovector[DEMirs$phenovector=="GroupA"])] %>% 
  as.data.frame() %>% 
  rownames_to_column("miR") %>% 
  gather(USI,log2.TMM.CPM, -miR) %>%
  mutate(miR=factor(miR,levels=highly.expressed$gene)) %>% 
  group_by(miR) %>%
  arrange(miR, desc(log2.TMM.CPM)) %>%
  ungroup() %>%
  mutate(USI=factor(USI, levels=unique(USI))) %>%
  spread(miR,log2.TMM.CPM) %>%
  # select(USI, all_of(highly.expressed$gene))


dim(highly.Expressed.TMM)

# write.csv(highly.Expressed.TMM,"TARGET_AML_CBFGLIS_HighlyExpressed_miRNAs_log2_TMM_CPM.csv", row.names = FALSE)
```


```{r}
sum(DEMirs.df$logFC >= 1)
sum(DEMirs.df$logFC <= 1)
```

probs = seq(0,1, length.out = 11)
seq(0,1, length.out = 11)

```{r}
quantile(DEMirs.df$logFC[DEMirs.df$logFC >= 1])
quantile(abs(DEMirs.df$logFC[DEMirs.df$logFC <= 1]))
```


```{r fig.height=5, fig.width=5}
extract_PCA(DEMirs)
extract_MDS(DEMirs)
```




##NBM

```{r}
# tiff("TARGET_AML_1031_CBGLIS_vs_NBM_DEMirs_Heatmap_10.29.18.tiff", width = 10, height = 10, units="in", res=600)

DEMirs.NBM <- twoGroups_DEMirs(expnData = mir.cts.1031, 
                           clinData = merged, 
                           col = "CBFA2T3.GLIS2", 
                           ref = "No",
                           RPM = FALSE,
                           BM = TRUE)
# dev.off()
```



```{r}
# save(DEMirs.NBM, file="TARGET_AML_1031_CBFGLIS_vs_NBM_DEMirs.RData")
DEMirs.NBM <- get(load("DEMirs/TARGET_AML_1031_CBFGLIS_vs_NBM_DEMirs.RData"))
table(DEMirs.NBM$phenovector)
```


```{r}
dim(DEMirs.NBM$DE$Voom$E) #569 mirs considered expressed. 
```


```{r}
DEMirs.NBM.df <- extract_DEMirs(DEMirs.NBM)

dim(DEMirs.NBM.df) #247

# write.csv(DEMirs.NBM.df, "DEMirs/TARGET_AML_1031_CBFGLIS_vs_NBM_DEMirs.csv",row.names = FALSE)
```

```{r}
sum(DEMirs.NBM.df$logFC >= 1)
sum(DEMirs.NBM.df$logFC <= 1)
```


#Visualize Mirs of Interest

```{r}
ggbox_dot_plots <- function(grouped.df,title,ylab=""){
  
  title <- unique(title)

 p <-  ggplot(grouped.df, aes(x=group, y=val, fill=group, color=group)) +
  labs(y=ylab, title=title) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(alpha=0.5) +
  scale_color_manual(values=c("GroupA"="red", "GroupB"="navy")) +
  theme_bw()
 
 
 return(p)
}
```


## miR-224/miR452 

```{r}
mir224_mir452_df <- RPM.1031 %>% 
  # cbind(., RPM.cells) %>%
  as.data.frame() %>%
  rownames_to_column("MIMAT") %>% 
  
  filter(grepl("-452\\.|-224\\.", MIMAT)) %>%
  gather(USI,RPM, -MIMAT) %>%
  mutate(Group=case_when(
    grepl("^Kas", USI) ~ "Kasumi", 
    grepl("MV4", USI) ~ "MV4", 
    grepl("^BM|^RO", USI) ~ "NBM", 
    TRUE ~ "AML")) %>% 
  left_join(., select(merged, USI,M7_AML, CBFA2T3.GLIS2), 
             by=c("USI")) %>%
  mutate(log2_RPM=log2(RPM+1)) %>% 
  
  mutate_at(vars(M7_AML:CBFA2T3.GLIS2), 
            ~case_when( Group == "Kasumi" ~ "Kasumi",
                        Group == "MV4" ~ "MV4",
                        Group == "NBM" ~ "NBM",
                        (is.na(.) & Group == "AML") ~ "Unknown", 
                        TRUE ~ .)) %>% 
  
  mutate(miR.v20=str_split_fixed(MIMAT, "\\.", n=2)[,1], 
         MIMAT=str_split_fixed(MIMAT, "\\.", n=2)[,2]) %>%
  left_join(., ID.map, by=c("MIMAT"="MIMAT.ID")) %>%
  dplyr::select(miR,miR.v20,MIMAT, everything())

                  

# head(mir224_mir452_df)
dim(mir224_mir452_df)
table(mir224_mir452_df$Group)
table(mir224_mir452_df$M7_AML,useNA = 'always')
```

```{r}
mir224_mir452_df %>% 
  filter(CBFA2T3.GLIS2 != "Unknown", M7_AML != "Unknown" ) %>%
  group_by(M7_AML,CBFA2T3.GLIS2, miR) %>% 
  summarise(N=n(), min=min(RPM), mean=mean(RPM),
            median=median(RPM), max=max(RPM))
```

```{r fig.height=6, fig.width=12}
mir.boxplot <- ggplot(filter(mir224_mir452_df,CBFA2T3.GLIS2 != "Unknown", 
              M7_AML != "Unknown", 
              !grepl("Kas|MV4",CBFA2T3.GLIS2)) , 
       aes(x=CBFA2T3.GLIS2, y=log2_RPM, fill=CBFA2T3.GLIS2)) +
  geom_boxplot(color="grey40") +
  # geom_violin(draw_quantiles = 0.5) +
  facet_wrap(~miR) +
  theme_classic() + 
  labs(title="miR-224 and miR-452 Expression", x="", y="log2 Reads per Million (RPM)") +
  scale_x_discrete(labels=c("NBM"="Normal\nMarrow", "No"="Fusion\nNegative", "Yes"="Fusion\nPositive")) + 
  scale_fill_manual(values=c("NBM"="grey80", "Yes"="red", "No"="navy")) +
  theme(strip.text = element_text(face="bold", size=18, color="black"), 
        text = element_text(size=18, color="black"), 
        axis.text = element_text(size=20, color = "black"), 
        legend.position = "none")

# mir.boxplot
# saveRDS(mir.boxplot, "TARGET_AML_CBFGLIS_miR224_miR452_boxplot.RDS")

# CairoPNG("TARGET_AML_CBFGLIS_miR224_miR452_boxplot.png",res=600,height = 6, width = 8,units="in")
mir.boxplot
# dev.off()

# ggsave(plot=mir.boxplot, "TARGET_AML_CBFGLIS_miR224_miR452_boxplot.png",
#        dpi=350, units = "in", height = 6, width = 8)
```


```{r}
high.miR <- mir224_mir452_df %>% 
  filter(CBFA2T3.GLIS2 != "Unknown", 
              M7_AML != "Unknown", 
              !grepl("Kas|MV4",CBFA2T3.GLIS2)) %>%
  # mutate(quartiles=quantcut(log2_RPM, q = 4, labels=paste0("q",1:4))) %>% 
  inner_join(., select(merged,-CBFA2T3.GLIS2), by="USI")

```


```{r fig.height=4}
# ggplot(filter(high.miR, MRD.at.end.of.course.1 != "Unknown"),
#        aes(x=MRD.at.end.of.course.1,y=log2_RPM, fill=MRD.at.end.of.course.1)) +
#   geom_boxplot() +
#   facet_wrap(~CBFA2T3.GLIS2)
  
ggplot(filter(high.miR, CR.status.at.end.of.course.1 != "Unknown"),
       aes(y=log2_RPM,
           x=CR.status.at.end.of.course.1,
           fill=CR.status.at.end.of.course.1)) +
  geom_boxplot() +
  facet_wrap(~CBFA2T3.GLIS2, scales="free_x")


# ggplot(filter(high.miR, Vital.Status != "Unknown"),
#        aes(y=log2_RPM,
#            x=Vital.Status,
#            fill=Vital.Status)) +
#   geom_boxplot() +
#   facet_wrap(~CBFA2T3.GLIS2, scales="free_x")
```

```{r fig.height=5, fig.width=5}
quantile(high.miR$log2_RPM)
quantile(log2(high.miR$MRD...at.end.of.course.1+1), na.rm = TRUE)
# quantile(high.miR$MRD...at.end.of.course.2, na.rm = TRUE)
plot(high.miR$log2_RPM,log2(high.miR$MRD...at.end.of.course.1+1))
cor.test(high.miR$log2_RPM,log2(high.miR$MRD...at.end.of.course.1+1))
```


## miRs uniquely expressed in CBFGLIS

```{r}
lowExpn <- DEMirs.df %>%
  filter(AveExpr < 0 & logFC >= 2)

dim(lowExpn) #15 strongly DE
range(lowExpn$adj.P.Val) #very sig
```

```{r}
Mirs.Sel <- DEMirs$DE$Voom$E %>%
  as.data.frame() %>%
  rownames_to_column("MIMAT") %>% 
  
  filter(MIMAT %in% lowExpn$gene | grepl("-452|-224", MIMAT)) %>%
  gather(var,val, -MIMAT) %>%
  
  inner_join(., data.frame(group=DEMirs$phenovector, USI=names(DEMirs$phenovector)), by=c("var"="USI")) %>%
  
  group_by(MIMAT) %>%
  do(plots=ggbox_dot_plots(., title=unique(.$MIMAT), ylab = "log2(CPM+0.5)"))

names(Mirs.Sel$plots) <- Mirs.Sel$MIMAT
```

```{r fig.height=3, fig.width=5}
# pdf("TARGET_AML_1031_CBFGLIS_vs_OtherAML_DEMIrs_HighInCBFGLIS_LowInOtherAML.pdf", height = 5, width = 7)
Mirs.Sel$plots$`hsa-mir-224.MIMAT0000281`
Mirs.Sel$plots$`hsa-mir-224.MIMAT0009198`
Mirs.Sel$plots$`hsa-mir-452.MIMAT0001635`
Mirs.Sel$plots$`hsa-mir-452.MIMAT0001636`
# dev.off()
```


#Visualize Volcanco plots 

```{r}
mirs <- rownames(DEMirs$DE$eBayesFit$coefficients) %>%
  str_split_fixed(., "\\.", n=2) %>%
  .[,1] %>%
  gsub("hsa-", "", .)

# volcanoplot(DEMirs$DE$eBayesFit, highlight = 5, names=mirs, pch=16, cex=1.75)

quantile(DEMirs.df$adj.P.Val, probs = seq(0,1,length.out = 11), type=5)
```

```{r}

mir.df <- data.frame(logFC=DEMirs$DE$eBayesFit$coefficients[,1],
                     pValue=DEMirs$DE$eBayesFit$p.value[,1],
                     FDR=p.adjust(DEMirs$DE$eBayesFit$p.value[,1], method="BH"), 
                     abs.t.stat=abs(DEMirs$DE$eBayesFit$t))

quantile( abs(mir.df$logFC[mir.df$FDR < 0.05]), probs = c(0,.75,.85,.90,.95,1))


mir.df <- mir.df %>%
  rownames_to_column("ID") %>%
  mutate(threshold = ifelse(logFC >= 1,"A", ifelse(logFC<=-1 , "B", "C"))) %>%
  mutate(mir.v20=str_split_fixed(ID, "\\.", n=2)[,1], 
         MIMAT=str_split_fixed(ID, "\\.", n=2)[,2]) %>%
  left_join(., ID.map, by=c("MIMAT"="MIMAT.ID")) %>%
  mutate(mir.label=gsub("hsa-", "", miR))  %>%
  
  rowwise() %>%
  mutate(mir.label=ifelse(FDR < 0.05 & abs(logFC) > 2.4, mir.label, "")) %>% 
  ungroup()


head(mir.df)
dim(mir.df)
# 684 - sum(grepl("^$", mir.df$mir.label))
```

g + geom_text(aes(x=gene1$logFC, y=-log10(gene1$P.Value),
                     label=gene1$ID, size=1.2), colour="black")


```{r fig.height=7, fig.width=10}
library(ggrepel)
require(circlize)
library(RColorBrewer)
```

```{r}
colorPal <- colorRampPalette(c("deepskyblue3", "deepskyblue2", "deepskyblue1","white","red1", "red2", "red3", "red4"))(n=20)
```

```{r fig.height=6.5, fig.width=9, warning=FALSE}
# tiff("CBFGLIS_vs_OtherAML_DeMirs_VolcanoPlot.tiff", height = 6.5, width = 10, res = 600, units = "in")
volcano <- ggplot(mir.df, aes(x=logFC, y=GroupA.GroupB, label = mir.label)) + 
  geom_point(aes(colour= logFC), size=5, alpha=1) +
  scale_color_gradientn(colours = colorPal, values=c(0,0.305,1), name="Log2 FC") +
  scale_x_continuous(limits = c(-5,9), breaks = c(-5,-3,-1,0,1,3,5,7,9)) +
  theme(plot.title = element_text(hjust = 0.5, size = 20),
                       panel.background = element_rect(fill="white"),
                       panel.grid.major = element_blank(),
                       panel.grid.minor = element_blank(),
                       panel.border = element_rect(color = "black", fill=NA),
                       axis.text = element_text(color = "black"),
                       axis.text.x = element_text(angle = 0,hjust=0.5,vjust = 0.5, size = 26),
                       axis.text.y = element_text(size = 25),
                       axis.title = element_text(size = 30),
                       plot.margin = margin(8,2,20,2), 
        legend.position = c(0.2,0.9),
        legend.direction = "horizontal",
        legend.text = element_text(size=18),
        legend.title = element_text(size=18, hjust = -0.5),
        legend.key.size =  unit(1.0, "cm")) + 
  labs(x="Log2 Fold-Change (FC)", y="Moderated t-Statistic", title="DE MiRNAs in CBFA2T3-GLIS2") +
    geom_text_repel(size=6, 
                    box.padding = unit(0.35, "lines"), 
                    point.padding = unit(0.3, "lines"))
 
# save(volcano, file="CBFGLIS_Volcanoplot.RData")
# dev.off()
```

 






#Session Information 

```{r}
sessionInfo()
```

